src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp

```diff
@@ -131,6 +131,13 @@ inline void PSPromotionManager::push_contents(oop obj) {
   }
 }
 
+inline void PSPromotionManager::push_contents_bounded(oop obj, HeapWord* left, HeapWord* right) {
+  if (!obj->klass()->is_typeArray_klass()) {
```

Comment:
I would probably put this check into `scan_obj_with_limit()` to also avoid the unnecessary prefetch and initialization of `PSPushContentsClosure`. `scan_obj_with_limit` seems to be the only caller anyway.